### PR TITLE
Deprecate the new auto tactic.

### PR DESCRIPTION
**Pull request by ppedrot**
It was not documented and not used in the CI either. It seems to have been introduced by e961ace3 but never really took off. Since this is responsible for a big chunk of code in Auto, it looks like it is worthwhile to deprecate it in order to remove it as soon as possible.

- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by ppedrot** [comment]
@silene this patch could even go into 8.14 in order not to delay the code removal it permits, but I would perfectly understand if you didn't want it to.

**Comment by ppedrot** [comment]
Also ping @mattam82 for an opinion on the matter.

**Comment by herbelin** [comment]
In passing, I'd be interested on knowing the plans of the team about automation in general. Do we mostly aim to maintain what exists or are there plans to get closer to our "dreams"?

**Comment by Zimmi48** [comment]
There's a lot of work going on by @mattam82 to improve automation supported by `typeclasses eauto`. AFAICT, the plans of the team are to focus the efforts there, by making programmable automation more and more controllable.

**Comment by herbelin** [comment]
Thanks for the answer.

More generally, I had typically general questions about things like:
- support for destructing tuples in hypotheses in the variants of eauto
- combination of rewriting and eauto
- combination of eauto and tactics such as congruence, or ring, or lia

I wanted to know if someone had projects in these directions.

**Comment by mattam82** [comment]
I'm fine with this deprecation. Indeed most of it is available through typeclasses eauto.

About Hugo's questions:

More generally, I had typically general questions about things like:

- support for destructing tuples in hypotheses in the variants of eauto

  One can have an Hint Extern for that. It should probably be marked as an invertible rule so we don't backtrack on it. Hint Extern When should be enough to handle that.
- combination of rewriting and eauto
  One can also use Hint Extern for that. If we want paramodulation then that's certainly another story, I'm not an expert on that.
- combination of eauto and tactics such as congruence, or ring, or lia
  I like the new approach of lia, I guess we should similarly parameterize congruence and ring to allow for mixing decision procedures. But again that's SMT-land that I'm no expert on. Maybe @MSoegtropIMC has ideas and use cases for this.

**Comment by herbelin** [comment]
Thanks for the answer. Then, is it a correct feeling that the `Hint Extern` technique is for relatively knowledgeable users? Or conversely, is there some collection of `Hint Extern` that should be made available out of the box without having to think (the same way as a collection of `Hint Resolve` and hint bases are provided - even if there are some mistakes in the building of these bases). [The question is quite naive. Automation is important but I know little about it. The question is more to have a feeling about where we go.]

**Comment by MSoegtropIMC** [comment]
Unfortunately I am a bit out of the `auto` proof business right now - I am mostly working on numerics these days, which requires different techniques, mostly specialized decision procedures I write in Ltac2 and sometimes in Mtac2. There is not much to gain from `auto` in this domain. My tactics typically make heavy use of reliable building blocks like `lia`, `ring`, `interval` and `gappa` and I very much appreciate the increase of robustness and flexibility in these.

Regarding hint extern: I used this extensively in cases a lemma needed guard conditions to decide if the application makes sense. I typically write a lemma and a tactic or a hint which uses it together (usually today this are tactics and not hints). What would make sense is to have a little dedicated language to describe such guard conditions which would work with some sort of fast pruning, e.g. which evaluates identical conditions for several hints only once.

**Comment by silene** [comment]
> this patch could even go into 8.14

There is a changelog entry, so backporting it would be painful.

**Comment by ppedrot** [comment]
@silene As you wish.

**Comment by Zimmi48** [comment]
@silene If that's the only issue, then you can ask @ppedrot to move the changelog entry to the appropriate location in the 8.14 release notes (that works only if the patch is then guaranteed to go in 8.14+beta1).

**Comment by ppedrot** [comment]
@silene ready to do that if you give your green light.

**Comment by silene** [comment]
Okay then.

**Comment by ppedrot** [comment]
@silene done. Can a maintainer quickly assign and merge once the CI is green, so as to make the life of our beloved RM easier?

File: user-contrib/Ltac2/Notations.v
Comment (herbelin) [inline user-contrib/Ltac2/Notations.v:582]: Do you really want a `+` or just `deprecated-ltac2-definition`.

Or maybe a section, so that the warnings takes back its original value?

**Review by herbelin — APPROVED** [review]
Quite canonical. See the comment though.

**Comment by herbelin** [comment]
Metacoq is failing everywhere, for apparently unrelated reasons.

Ping me, if I forgot to merge (actually it will be ok if @silene or @SkySkimmer happens to pass here and merge it).

File: user-contrib/Ltac2/Notations.v
Comment (ppedrot) [reply user-contrib/Ltac2/Notations.v:582]: These various options are functionally equivalent in the current context, and the current code is short.

**Comment by ppedrot** [comment]
metacoq is broken on master, I am trying to fix it with input from some of its devs but it ain't easy. I think this PR can be safely merged. The only important thing is the test-suite and the linter, since the PR removing the tactic had a green CI two days ago.

File: doc/sphinx/changes.rst
Comment (Zimmi48) [inline doc/sphinx/changes.rst:273]: Please insert after the last **Deprecated** and before the first **Added** in this category.

File: user-contrib/Ltac2/Notations.v
Comment (herbelin) [reply user-contrib/Ltac2/Notations.v:582]: > These various options are functionally equivalent in the current context

You mean because its scope is only the file anyway?

>  and the current code is short.

What do you mean?

Comment (ppedrot) [reply user-contrib/Ltac2/Notations.v:582]: Yes, the scope is only the remainder of the file. It's short compared to adding sections and relying on auto discharge.

Comment (herbelin) [reply user-contrib/Ltac2/Notations.v:582]: But removing the `+` would be even shorter! Anyway, this is nitpicking.

File: doc/sphinx/changes.rst
Comment (herbelin) [reply doc/sphinx/changes.rst:273]: @Zimmi48: Concretely, how can this be done?

Comment (Zimmi48) [reply doc/sphinx/changes.rst:273]: I thought my request was crystal clear but since it isn't, I'll take care of pushing the change.

Comment (Zimmi48) [reply doc/sphinx/changes.rst]: Done.

Comment (herbelin) [reply doc/sphinx/changes.rst]: Ah, ok. I did not realize that @ppedrot had modified `changes.rst` directly.

**Comment by ppedrot** [comment]
@herbelin this is ready to merge.

**Comment by herbelin** [comment]
@ppedrot: Would you mind if I push my `+` nitpicking on top of the PR before merging?

**Comment by ppedrot** [comment]
@herbelin go ahead, but this needs to be merged quickly since this must go in 8.14.

**Comment by ppedrot** [comment]
(@herbelin Please don't put a section though, this is complexifying the code and might have weird side-effects.)

**Comment by ppedrot** [comment]
@herbelin you messed up the rebase it seems.

**Comment by herbelin** [comment]
UniMath is failing elsewhere too. I suppose we don't have to care.

@coqbot: merge now

**Pull request merged**
